### DISCUSSION TITLE
No response to LS `textDocument/documentSymbol` request

### DISCUSSION THREAD

**Discussion by darius-sas** [microsoft/pyright#3863] [Q&A] [answered]
I'm trying to integrate with Pyright so that I can analyze Python files with my research tool (written in Java).
To do so, I'm using pyright-langserver to retrieve the information of each file via a LSP `textDocument/documentSymbol` request.
Unfortunately, I cannot get the server to respond to any request after I send the `initialize` and `initialized` requests. 

Here's the logging messages I get back from the server (I'm trying to analyse a sample python project called _python-telegram-bot_):
```java
14:16:00 [pool-2-thread-1] INFO  c.a.l.Logger - Pyright language server 1.1.268 starting
14:16:00 [pool-2-thread-1] INFO  c.a.l.Logger - Server root directory: /tmp/pyright-python-langserver.fenn/node_modules/pyright/dist/
14:16:00 [pool-2-thread-1] INFO  c.a.l.Logger - Starting service instance "/path/to/python-telegram-bot/"
```

As soon as I send a `textDocument/documentSymbol` the server goes silent and I don't receive a response (my client hangs indefinetely for a response). Is there something I'm doing wrong? 
The command line I start the LS is the following: `pyright-langserver --stdio`
This is my initialization request:
```json
{
  "jsonrpc": "2.0",
  "id": "1",
  "method": "initialize",
  "params": {
    "workDoneToken": "initialize-to-project",
    "processId": null,
    "rootPath": "file:///path/to/python-telegram-bot/",
    "rootUri": "file:///path/to/python-telegram-bot/",
    "capabilities": {
      "workspace": {},
      "textDocument": {
        "documentSymbol": {
          "symbolKind": {
            "valueSet": [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]
          },
          "hierarchicalDocumentSymbolSupport": true,
          "dynamicRegistration": true
        }
      }
    },
    "workspaceFolders": [
      {
        "uri": "file:///path/to/python-telegram-bot/"
      }
    ]
  }
}
```

My LS client implementation works well with other language servers like [OmniSharp](https://github.com/OmniSharp/omnisharp-roslyn), [python-language-server](https://github.com/python-lsp/python-lsp-server), and [EclipseJDTLS](https://github.com/eclipse/eclipse.jdt.ls), so I think it's a configuration problem with pyright.
Note that I prefer not using py-lsp as it seems to not be supporting DocumentSymbol and is much less precise than pyright.

I tried to find a solution in other issues and via Google but found nothing. Hopefully you guys can help, I'm kinda DESPERATE!!

**Comment by erictraut** [comment]
It sounds like you will need to do some debugging to dig into the problem further. Please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/build-debug.md).

**Comment by darius-sas** [comment, marked as answer]
Thanks @erictraut for the quick reply!

It took a while (I'm not a JS/TS developer), but after some digging and debugging I managed to find the problem. In short it was that Pyright's LS expects a `workspace/didChangeWorkspaceFolders` notification being sent after an `initialized` is sent by the client, otherwise it does not update its workspace list. 
Here's the [code responsible for that](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/languageServerBase.ts#L631).
I believe, however, that this should not be the default behaviour and that the workspace should be automatically updated after the `initialized` request is received, otherwise it seems to deviate from the LSP specification. Let me know what you think.

**Reply by erictraut** [reply]
OK, that diagnosis sounds plausible to me. I'd be somewhat reluctant to accept a code change here because it has a chance of introducing regressions that could affect other consumers of the pyright LSP. It sounds like you have a straightforward workaround for your use case.

**Reply by SoryEscobar** [reply]
Hi @darius-sas! Im running into the same issue here. Im interested in using the documentSymbol capabilities for pyright over nvim-lsp and I haven't found the solution yet, by your comment seems like you did find it, how ever im not quite sure to understand how did you implemented that. Could you please expand on this? If you have configuration files or something like that would be great!

**Reply by darius-sas** [reply]
Hi @SoryEscobar, my problem arised because I was programmatically implementing a LSP client for pyright and my client was not behaving as the server would expect (which was slightly different than the LS protocol mentioned). My solution to the problem was to send a `workspace/didChangeWorkspaceFolders` request to the server right after I receive an `initialized` response.
This is the kotlin code I wrote for the job to get you a better idea
```kotlin
initializeResult = server!!.initialize(initializeParams).join()
waitUntilReady()
initialized()
didChangeWorkspaceFolders(workspaces, listOf())
````

**Reply by SoryEscobar** [reply]
Thanks, @darius-sas for the response! I'll review your provided solution and validate if it applies to my issue. Regards!

**Comment by darius-sas** [comment]
I leave here for the search engine bots the steps you want to do to debug the language server with VS Code:

1. Install [ts-node](https://www.npmjs.com/package/ts-node)
2. run `npm run webpack` in the main directory of the project. 
3. Add this run configuration to VS Code `launch.json` file
```json
        {
            "name": "PyrightLangServer Debug",
            "type": "node",
            "request": "launch",
            "runtimeExecutable": "node",
            "runtimeArgs": ["--nolazy", "-r", "ts-node/register"],
      
            "args": ["packages/pyright-internal/src/nodeMain.ts", "--socket", "5643"], // Change the args according to your needs
            
            "cwd": "${workspaceRoot}",
            "internalConsoleOptions": "openOnSessionStart",
            "skipFiles": ["<node_internals>/**", "node_modules/**"]
        },
```
Credits to this [gist](https://gist.github.com/cecilemuller/2963155d0f249c1544289b78a1cdd695) for the original config

4. Add the following call to main `main()` at the end of the `packages/pyright-internal/src/nodeMain.ts` file
5. Run from the debugger window of VS Code `PyrightLangServer Debug`

Hopefully this is helpful to someone.
